crypto/params.c

```diff
@@ -0,0 +1,369 @@
+
+    return 1;
+}
+
+#define PARAM_FIXED(name, type, PARAM) \
+    int OSSL_PARAM_set_##name(OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return set_fixed(p, key, val, PARAM); \
+    } \
+    int OSSL_PARAM_reserve_##name(OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return reserve_fixed(p, key, val, PARAM); \
+    } \
+    int OSSL_PARAM_get_##name(OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return get_fixed(p, key, val, PARAM); \
+    } \
+    int OSSL_PARAM_return_##name(OSSL_PARAM *p, const char *key, type val) \
+    { \
+        return return_fixed(p, key, &val, PARAM); \
+    }
+
+PARAM_FIXED(int, int, OSSL_PARAM_INT)
+PARAM_FIXED(uint, unsigned int, OSSL_PARAM_UINT)
+PARAM_FIXED(int64, int64_t, OSSL_PARAM_INT64)
+PARAM_FIXED(uint64, uint64_t, OSSL_PARAM_UINT64)
+PARAM_FIXED(long, long int, OSSL_PARAM_LONG)
+PARAM_FIXED(ulong, unsigned long, OSSL_PARAM_ULONG)
+PARAM_FIXED(size_t, size_t, OSSL_PARAM_SIZET)
+PARAM_FIXED(double, double, OSSL_PARAM_DOUBLE)
+PARAM_FIXED(pointer, void*, OSSL_PARAM_POINTER)
+
+int OSSL_PARAM_set_bignum(OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    if (key != NULL && (p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    if (p->data_type != OSSL_PARAM_BIGNUM) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_BIGNUM, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    p->buffer = *val;
+    return 1;
+}
+
+int OSSL_PARAM_reserve_bignum(OSSL_PARAM *p, const char *key,
+                              void *buffer, size_t bufsize)
+{
+    return reserve_ptrsize(p, key, buffer, bufsize);
+}
+
+int OSSL_PARAM_get_bignum(OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    if (key != NULL && (p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    *val = *(BIGNUM **)p->buffer;
+    return 1;
+}
+
+int OSSL_PARAM_return_bignum(OSSL_PARAM *p, const char *key, BIGNUM *val)
+{
+    size_t bytes = (size_t)BN_num_bytes(val);
+    int r;
+
+    if (key != NULL && (p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_RETURN_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM
+            || BN_is_negative(val)) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_RETURN_BIGNUM, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    if (p->size < bytes
+            || (r = BN_bn2nativepad(val, p->buffer, bytes)) < 0)
+        return 0;
+    p->used = r;
```

Comment:
This should happen before the check of `p->size`, to give the caller a fighting chance to resize appropriately.

Reply:
I understand the rationale, but ```r``` will be -1 if there's no size<bytes.  I changed things around to include an output "guess" but I'm not thrilled with that.  New commit pushed.